### DISCUSSION TITLE
Does Capsule have support to reduce noisy neighbours.

### DISCUSSION THREAD

**Discussion by saikumar0412** [projectcapsule/capsule#1302] [Q&A]
Is capsule having the feature to reduce resources automatically? when a tenant quota is under utilized? which can help to reduce costs?

**Comment by oliverbaehler** [comment]
So wouldnt that result into the quota boundaries always being equal to actual usage?
For me personally you have to handle this on workload level via autoscaling not via resourcequotas unless i am misunderstanding your use-case

**Reply by saikumar0412** [reply]
Workload level scaling is not an issue we can always do that either using HPA or KEDA but When it comes to a situation tenant one have been assigned some resources quota and it only using 20-30% of it(not even 100%) which can cause cost increase at the end. Do you recommaned/ have any solution for this problem.

**Reply by prometherion** [reply]
Joining the party later.

Kubernetes offers a `scale` subresource for CRDs mostly based on replicas, and we can't bend it for Capsule. The resource quota resizing is a bit complicated since it should be addressed with several weights and conditions which Capsule doesn't offer.

My suggestion here would be to rely on a custom controller with your logic embedded in it (or programmatically assigned) which performs the changes on the Tenant Custom Resources.

As Capsule is missing maintainers and all the burden is on top of CLASTIX and Peak Scale organizations, we're happy to receive contributions, or financial support in implementing these kinds of additions.

**Reply by oliverbaehler** [reply]
Here's the thing, i am also trying to solve the quota racing condition for quotas (#49). There we encounter the problem, that resourcequotas spread across multiple namespaces refeering to one tenant quota can be overprovisioned, if an operation is executed in parallel (eg. total is services/count: 3, in each namespace you could then create 3 services, leading to a possible overprovision of hard * namespaces).

Here we have the problem, that even if we would block resourcequota status updates and wait until the actual quantity was added to the total, the resources have already been scheduled. The reason for that, is that the status for resourcequotas is essentially consistent, but what really matters at that moment is the hard spec (see this response from a maintainer https://github.com/kubernetes/kubernetes/issues/123434#issuecomment-1964920277)

So the cleanest way imho would be doing the allocation of available resources in reverse:

1. All resourcequotas get for their hard spec 0
2. New resource is request
3. Controller gives the requested resources to the quota (by adding it to the total and updating the hard)

This way it's only possible to scheduled "ordered". In conclution this would also downscale the resourcequota when the resources are no longer needed. 

But there's som, problems with this approach as well:

*  if you eg. schedule a pod and the quota is count/0 there's no admission call on the resourcequota, which would be the easiest.  So we would need to find a way to know, there's something new requesting resources. The only indication from what i know of is that we get an Event, which we can intercept with admission, regarding quotaoverprovision.
* If you eg update the resource quota that a pod now has space, it takes some time until that's registered and actually scheduled (just tested it for pods). I guess the timing depends on the kube-controller-manager flag `--concurrent-resource-quota-syncs ` and/or `--resource-quota-sync-period`

So it's really really difficult to icnrease quotas by the resources which are actually requested, especially the adding new resources process is where the performance would take a heavy hit.

Just what I was thinking about, happy for any other ideas.
